In [1]:
import os
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import mnist

2025-05-28 21:17:05.431820: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:467] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
E0000 00:00:1748485025.695296   35957 cuda_dnn.cc:8579] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
E0000 00:00:1748485025.793381   35957 cuda_blas.cc:1407] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
W0000 00:00:1748485026.276907   35957 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1748485026.276940   35957 computation_placer.cc:177] computation placer already registered. Please check linkage and avoid linking the same target more than once.
W0000 00:00:1748485026.276942   35957 computation_placer.cc:177] computation placer alr

In [2]:
# load and prepare data for training
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(-1, 28 * 28).astype("float32") / 255.0
x_test = x_test.reshape(-1, 28 * 28).astype("float32") / 255.0

## Dense Feed Forward Neural Networks (DFNNs)  also known as multi-layer perceptrons(MLPs)
* well suited for inferencing from tabular data

### Construct DFNN model using Sequential API  
* simple interface
* limited functionality

In [3]:
# Sequential API (Very convenient, not very flexible)
modelDFNN1 = keras.Sequential(
    [
        keras.Input(shape=(784,)),
        layers.Dense(512, activation="relu"),
        layers.Dense(256, activation="relu"),
        layers.Dense(10),
    ]
)

modelDFNN1 = keras.Sequential()
modelDFNN1.add(keras.Input(shape=(784,)))
modelDFNN1.add(layers.Dense(512, activation="relu"))
modelDFNN1.add(layers.Dense(256, activation="relu", name="my_layer"))
modelDFNN1.add(layers.Dense(10))

I0000 00:00:1748485036.887906   35957 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 15494 MB memory:  -> device: 0, name: Tesla P100-PCIE-16GB, pci bus id: 0000:02:00.0, compute capability: 6.0
I0000 00:00:1748485036.888990   35957 gpu_device.cc:2019] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 15506 MB memory:  -> device: 1, name: Tesla P100-PCIE-16GB, pci bus id: 0000:82:00.0, compute capability: 6.0


In [4]:
# ToDo #1: List what objects the modelDFNN1 is comprised of (built from)
print(modelDFNN1.summary())


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 512)            │       401,920 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ my_layer (Dense)                │ (None, 256)            │       131,328 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 10)             │         2,570 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 535,818 (2.04 MB)

 Trainable params: 535,818 (2.04 MB)

 Non-trainable params: 0 (0.00 B)

None


In [5]:
modelDFNN1.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=False),
    optimizer=keras.optimizers.Adam(learning_rate=0.001),
    metrics=["accuracy"],
)

In [6]:
modelDFNN1.fit(x_train, y_train, batch_size=32, epochs=5, verbose=2)
modelDFNN1.evaluate(x_test, y_test, batch_size=32, verbose=2)

Epoch 1/5


I0000 00:00:1748485040.174435   54624 service.cc:152] XLA service 0x2aaea0002c60 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1748485040.174819   54624 service.cc:160]   StreamExecutor device (0): Tesla P100-PCIE-16GB, Compute Capability 6.0
I0000 00:00:1748485040.174825   54624 service.cc:160]   StreamExecutor device (1): Tesla P100-PCIE-16GB, Compute Capability 6.0
2025-05-28 21:17:20.287561: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:269] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
I0000 00:00:1748485040.544611   54624 cuda_dnn.cc:529] Loaded cuDNN version 90300
I0000 00:00:1748485041.370527   54624 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.


1875/1875 - 5s - 3ms/step - accuracy: 0.1324 - loss: 2.3317
Epoch 2/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 3/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 4/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 5/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
313/313 - 1s - 4ms/step - accuracy: 0.1060 - loss: 2.3026


[2.30259108543396, 0.10599999874830246]

In [7]:
# ToDo #2: What train and test accuracy do you obtain when running for 5 epochs ?  

modelDFNN1.fit(x_train, y_train, batch_size=32, epochs=5, verbose=2)
modelDFNN1.evaluate(x_test, y_test, batch_size=32, verbose=2)


Epoch 1/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 2/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 3/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 4/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 5/5
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
313/313 - 0s - 2ms/step - accuracy: 0.1060 - loss: 2.3026


[2.30259108543396, 0.10599999874830246]

In [8]:
# ToDo #3: Now write code (add code cells to this notebook) to train the model for 20 additional epochs. What train and test accuracy do you obtain? 
modelDFNN1.fit(x_train, y_train, batch_size=32, epochs=25, verbose=2)
modelDFNN1.evaluate(x_test, y_test, batch_size=32, verbose=2)



Epoch 1/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 2/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 3/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 4/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 5/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 6/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 7/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 8/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 9/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 10/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 11/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 12/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 13/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
Epoch 14/25
1875/1875 - 3s - 2ms/step - accuracy: 0.1056 - loss: 2.3026
E

[2.30259108543396, 0.10599999874830246]

### Construct DFNN model using Functional API  
* slightly more complex interface
* additional functionality

In [9]:
# ToDo #4: Now practice using the OO functional API of keras by constructing your own functional model, modelDFNN2.
# To simplify the task, create this model so that it uses the same architecture as the sequential model.
# For the final Dense layer use activation="softmax"


In [11]:
# Functional API (A bit more flexible)
inputs = keras.Input(shape=(784,))
x = layers.Dense(512, activation="relu", name="first_layer")(inputs)
x = layers.Dense(256, activation="relu", name="second_layer")(x)
outputs = layers.Dense(10, activation="softmax")(x)
modelDFNN2 = keras.Model(inputs=inputs, outputs=outputs)





In [ ]:
# Now to gain experience invoking methods on objects you have created,
# write code to compile(), fit() your new model on the training data. 
# Then write code to evaluate your new model on the test data. 


In [13]:
modelDFNN2.compile(
    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=False),
    optimizer=keras.optimizers.Adam(learning_rate=0.001),
    metrics=["accuracy"],
)

In [14]:
modelDFNN2.fit(x_train, y_train, batch_size=32, epochs=5, verbose=2)
modelDFNN2.evaluate(x_test, y_test, batch_size=32, verbose=2)

Epoch 1/5
1875/1875 - 4s - 2ms/step - accuracy: 0.9417 - loss: 0.1887
Epoch 2/5
1875/1875 - 3s - 2ms/step - accuracy: 0.9757 - loss: 0.0786
Epoch 3/5
1875/1875 - 3s - 2ms/step - accuracy: 0.9824 - loss: 0.0555
Epoch 4/5
1875/1875 - 3s - 2ms/step - accuracy: 0.9869 - loss: 0.0396
Epoch 5/5
1875/1875 - 3s - 2ms/step - accuracy: 0.9904 - loss: 0.0303
313/313 - 1s - 4ms/step - accuracy: 0.9782 - loss: 0.0792


[0.07915882021188736, 0.9782000184059143]

In [ ]:
# ToDo #5. Write down what train and test accuracy do you observe? How does it compare to the sequential model?

accuracy is 0.9782000184059143 which is much higher than sequenttial model

In [ ]:
# For fun: explore the architecture space by adding or removing layers from your model, retrain from scratch. 
# Learn how that impacts model performance. State of the art accuracy is around 99.97% . 
# You do not need to achieve that (nor is that expected here) but to give you an idea.